### DISCUSSION TITLE
Suggestions for developing and testing an existing installer to add a "Burn" that installs dependencies?

### DISCUSSION THREAD

**Discussion by cinderblock** [wixtoolset/issues#8660] [Questions]
Hello,

I'm looking into if it would be possible to use more of WiX's features to install optional dependencies when installing a program. Specifically, I'm looking into [updating Node.js's installer to not depend on Chocolatey](https://github.com/nodejs/node/issues/51905).

StackOverflow suggests I should use a "Burn". I've found the docs for creating Burn bundles - but I haven't found complete examples to follow. It seems I'm missing some fundamental ideas on how the build chains work.

I'm following what happens in [vcbuild.bat](https://github.com/nodejs/node/blob/main/vcbuild.bat) but it seems a bit heavy to develop with. If nothing else, I can push forward with that / GitHub Actions to build everything. But this will be a painfully slow development cycle.

Can anyone here get me pointed in the the right direction or short circuit any necessary learnings?

**Comment by robmen** [comment]
It is probably easier to answer specific questions. You can find a working example of a bundle in the WiX codebase at: https://github.com/wixtoolset/wix/tree/main/src/setup/WixAdditionalTools

**Reply by cinderblock** [reply]
Yeah, unfortunately I don't have enough understanding to have specific questions yet. II hadn't found this example. Thank you.

**Reply by cinderblock** [reply]
Ok. One question that may drive other decisions - would a [burn] bundle need to include all the optional assets to be installed or can a burn download additional binaries from the internet?

**Reply by bevanweiss** [reply]
It depends what you mean by 'download additional binaries...', if you mean binaries that you have included in the bundle definition, then yes, it can source them from a downloadURL.
If you mean that it goes to an API of some form, and retrieves a list of things that it presents to the user as options to download, then the WiX Standard bootstrapper doesn't have that out of the box (that I'm aware of), you'd need to create a custom bootstrapper.

You really should read through the Schema to see what options are available.  Your question would have been answered there.

**Reply by cinderblock** [reply]
Thank you for your helpful suggestions for solving my questions on this discussion!

I understand your confusion about the possible double meaning of "download additional binaries". Basically, all possibilities are potentially on the table. One simple question is determining if the downloaded installer bundle for Node.js would need to grow significantly (many megabytes) to also include the VCTools & Python installers. It sounds like maybe that's the "standard" way to do it. That may be a non-starter for a change to the Node.js installer. Clearly the current Chocolatey way downloads the tools separately so they aren't bundled in the Node.js installer. The Winget solution would be very similar. Of course this causes problems for offline installers - but you could always have alternate installer bundles for them. Maybe this would open the door to a new single unified Node.js installer that downloads the appropriate binaries (x86/x64/ARM64) for the current system? That would probably be a bigger change for the Node.js project that would need some more buy-in from concerned parties. But maybe this would be desired? That's well beyond my concerns.

Getting back on task...

I see [18 different Schemas](https://wixtoolset.org/docs/schema/) listed in the WiX Toolset package. I suspect you're talking about the [Wxs Schema](https://wixtoolset.org/docs/schema/wxs/), as the only seemingly relevant one with many options. That has 297 Elements to read about and more Types. Which Element has the feature I'm looking for? Or, are you saying none of them do? That would save me some trouble. What docs am I not seeing?

What's the difference between a Burn and a custom bootstrapper? Is a "Burn" the "WiX Standard bootstrapper"? I see that documentation for a v3 bootstrapper doesn't have an equivalent in v4/v5. Yay. Another rabbit hole to go down!

**Reply by bevanweiss** [reply]
Downloading packages is quite common, it's how most pre-requisites such as .NET are done in online intended installers.
It's possible to cache the files 'online' to allow for offline installation also (think of how Visual Studio does it.. it's very custom, but uses, at least used to, WiX for its installation mechanisms, as do many other Microsoft products).

The `Bal` schema and in particular around Bundles and Packages should be your first point of call.
You don't need to read every single item, but having a concept of the options available (in terms of elements) would benefit you.

**Reply by cinderblock** [reply]
Thank you for the specific recommendation to look at the `Bal` schema. My initial cursory look suggests in have more Windows app stuff to get more familiar with.

Yeah, I suspect the VS installer, as an example, is quite a bit more complicated than should be necessary for this simple use case.

I suspect Node.js won't want to include a much more complicated and custom installer. I'm not very worried about caching files. I'd think providing an "offline" and an "online" installer would be sufficient to make download sizes smaller for users that don't need everything while still providing a "complete" offline installer with all options available.

**Reply by robmen** [reply]
You don't need to understand all 297 elements. I do recommend you take a look at the Bundle that ships as part of the WiX Toolset today in: https://github.com/wixtoolset/wix/tree/main/src/setup/WixAdditionalTools

There are 10 unique elements in that fully functional Bundle's .wxs file. You'll probably need to add an `ExePackage` (or maybe even `BundlePackage`, since IIRC Python installs using a WiX-based Bundle), but reading the documentation for each of those 10 elements should get you very far along.

After you get the install/uninstall of the packages you want, I expect you'll have a solid foundation for the more advanced topics after that.

**Reply by robmen** [reply]
PS: If you use DownloadUrl attributes to build a bundle that downloads some of its packages from the internet (instead of carrying them with/embedded in your bundle) to minimize the size, you can still get an "offline installer" by using the `-layout` switch. That switch tells your bundle .exe to download all the external data into a "local layout". A bundle .exe launched from a local layout has all its files and will not need network access to install, perfect for offline scenarios.

**Reply by cinderblock** [reply]
Thank you @robmen. I think this gets me pointed in a good direction.

**Comment by bevanweiss** [comment]
It's possible that Burn might be a suitable replacement for the Node.js usage of chocolatey, but you probably want to do some further investigation into how chocolatey currently installs some of these things.
A Burn chain with each of these as an EXE/MSI Package would probably do it if installation of all was required.  But trying to sort out the cross dependencies from existing chocolatey authoring might be painful if there's supposed to be some selection between them.

The list of chocolatey installation items from one of the secondary links in the node.js issue are below... knowing which KB to install for various supported platforms seems a bit dicey.  Are they all valid for WinXP through Win11, and Windows Server 2000 through Windows Server 2022.. I wouldn't expect so..
- kb2919355
- kb2919442
- kb2999226
- kb3033929
- kb3035131
- python3
- dotnetfx
- vcredist140
- visualstudio2019-workload-vctools
- visualstudio2019buildtools
- visualstudio-installer

chocolatey also states that it can keep them all up to date.  I'm not entirely sure what that actually means.  But I assume that for the likes of Python, it will go and fetch the latest version if an update is done.  For that kind of behaviour, you'll start to venture into the custom bootstrapper realm, but it's all based around similar concepts.
https://wixtoolset.org/docs/tools/burn/

**Reply by cinderblock** [reply]
I expect the Python and VCTools installers will install their necessary dependencies.

Chocolatey is a package manager and, like most package managers, has a single command to update all installed packages. Winget has this functionality as well. However, many users seem to not want some extra package manager running updates in the background. I'm indifferent but it feels like the default should be to follow the "Windows" standard and not update unless specifically told to by the user.

**Comment by rsmart8452** [comment]
If you decide WiX is the way to go, here's a tutorial with practical examples.

https://github.com/rsmart8452/Wix4BurnTutorial

**Reply by cinderblock** [reply]
Thank you. This example looks helpful.
